### PR TITLE
Improve allocation of properties

### PR DESCRIPTION
**Pull request by funny-falcon-at-joomcode**
```
benchmark                         old ns/op     new ns/op     delta
BenchmarkAllocProperty/props0     100           83.0          -17.00%
BenchmarkAllocProperty/props1     319           172           -46.08%
BenchmarkAllocProperty/props2     628           292           -53.50%
BenchmarkAllocProperty/props3     1033          443           -57.12%
BenchmarkAllocProperty/props4     1485          624           -57.98%
BenchmarkAllocProperty/props5     2012          828           -58.85%
BenchmarkAllocProperty/props6     2573          1062          -58.73%
BenchmarkAllocProperty/props7     3201          1375          -57.04%
BenchmarkAllocProperty/props8     3829          1591          -58.45%

benchmark                         old allocs     new allocs     delta
BenchmarkAllocProperty/props0     2              1              -50.00%
BenchmarkAllocProperty/props1     5              3              -40.00%
BenchmarkAllocProperty/props2     8              6              -25.00%
BenchmarkAllocProperty/props3     11             10             -9.09%
BenchmarkAllocProperty/props4     14             14             +0.00%
BenchmarkAllocProperty/props5     17             19             +11.76%
BenchmarkAllocProperty/props6     20             25             +25.00%
BenchmarkAllocProperty/props7     23             32             +39.13%
BenchmarkAllocProperty/props8     26             37             +42.31%

benchmark                         old bytes     new bytes     delta
BenchmarkAllocProperty/props0     176           96            -45.45%
BenchmarkAllocProperty/props1     672           240           -64.29%
BenchmarkAllocProperty/props2     1168          432           -63.01%
BenchmarkAllocProperty/props3     1664          672           -59.62%
BenchmarkAllocProperty/props4     2160          912           -57.78%
BenchmarkAllocProperty/props5     2656          1200          -54.82%
BenchmarkAllocProperty/props6     3152          1536          -51.27%
BenchmarkAllocProperty/props7     3648          1920          -47.37%
BenchmarkAllocProperty/props8     4144          2208          -46.72%
```

### CODE REVIEW

**Comment by funny-falcon-at-joomcode** [comment]
Added commit "Move "underlying errors" into property".
New comparison:
```
benchmark                         old ns/op     new ns/op     delta
BenchmarkAllocProperty/props0     100           67.7          -32.30%
BenchmarkAllocProperty/props1     319           148           -53.61%
BenchmarkAllocProperty/props2     628           262           -58.28%
BenchmarkAllocProperty/props3     1033          375           -63.70%
BenchmarkAllocProperty/props4     1485          530           -64.31%
BenchmarkAllocProperty/props5     2012          713           -64.56%
BenchmarkAllocProperty/props6     2573          969           -62.34%
BenchmarkAllocProperty/props7     3201          1099          -65.67%
BenchmarkAllocProperty/props8     3829          1297          -66.13%

benchmark                         old allocs     new allocs     delta
BenchmarkAllocProperty/props0     2              1              -50.00%
BenchmarkAllocProperty/props1     5              3              -40.00%
BenchmarkAllocProperty/props2     8              6              -25.00%
BenchmarkAllocProperty/props3     11             9              -18.18%
BenchmarkAllocProperty/props4     14             13             -7.14%
BenchmarkAllocProperty/props5     17             18             +5.88%
BenchmarkAllocProperty/props6     20             24             +20.00%
BenchmarkAllocProperty/props7     23             28             +21.74%
BenchmarkAllocProperty/props8     26             33             +26.92%

benchmark                         old bytes     new bytes     delta
BenchmarkAllocProperty/props0     176           64            -63.64%
BenchmarkAllocProperty/props1     672           176           -73.81%
BenchmarkAllocProperty/props2     1168          336           -71.23%
BenchmarkAllocProperty/props3     1664          496           -70.19%
BenchmarkAllocProperty/props4     2160          704           -67.41%
BenchmarkAllocProperty/props5     2656          960           -63.86%
BenchmarkAllocProperty/props6     3152          1264          -59.90%
BenchmarkAllocProperty/props7     3648          1472          -59.65%
BenchmarkAllocProperty/props8     4144          1728          -58.30%
```

**Comment by funny-falcon-at-joomcode** [comment]
And made properties implemented as linked list:
```
benchmark                         old ns/op     new ns/op     delta
BenchmarkAllocProperty/props0     100           68.2          -31.80%
BenchmarkAllocProperty/props1     319           135           -57.68%
BenchmarkAllocProperty/props2     628           202           -67.83%
BenchmarkAllocProperty/props3     1033          275           -73.38%
BenchmarkAllocProperty/props4     1485          341           -77.04%
BenchmarkAllocProperty/props5     2012          408           -79.72%
BenchmarkAllocProperty/props6     2573          474           -81.58%
BenchmarkAllocProperty/props7     3201          540           -83.13%
BenchmarkAllocProperty/props8     3829          607           -84.15%

benchmark                         old allocs     new allocs     delta
BenchmarkAllocProperty/props0     2              1              -50.00%
BenchmarkAllocProperty/props1     5              3              -40.00%
BenchmarkAllocProperty/props2     8              5              -37.50%
BenchmarkAllocProperty/props3     11             7              -36.36%
BenchmarkAllocProperty/props4     14             9              -35.71%
BenchmarkAllocProperty/props5     17             11             -35.29%
BenchmarkAllocProperty/props6     20             13             -35.00%
BenchmarkAllocProperty/props7     23             15             -34.78%
BenchmarkAllocProperty/props8     26             17             -34.62%

benchmark                         old bytes     new bytes     delta
BenchmarkAllocProperty/props0     176           64            -63.64%
BenchmarkAllocProperty/props1     672           160           -76.19%
BenchmarkAllocProperty/props2     1168          256           -78.08%
BenchmarkAllocProperty/props3     1664          352           -78.85%
BenchmarkAllocProperty/props4     2160          448           -79.26%
BenchmarkAllocProperty/props5     2656          544           -79.52%
BenchmarkAllocProperty/props6     3152          640           -79.70%
BenchmarkAllocProperty/props7     3648          736           -79.82%
BenchmarkAllocProperty/props8     4144          832           -79.92%
```

File: builder.go [hunk 1]
```diff
@@ -12,21 +12,20 @@ type ErrorBuilder struct {
...
 	message       string
 	cause         error
-	underlying    []error
 	mode          callStackBuildMode
 	isTransparent bool
 }
 
 // NewErrorBuilder creates error builder from an existing error type.
-func NewErrorBuilder(t *Type) *ErrorBuilder {
+func NewErrorBuilder(t *Type) ErrorBuilder {
```
Comment (PeterIvanov) [inline builder.go]: ErrorBuilder is routinely copied several times, are you sure that value receiver is an improvement here?

File: error.go [hunk 2]
```diff
@@ -27,39 +27,45 @@ var _ fmt.Formatter = (*Error)(nil)
...
+}
 
-	if errorCopy.properties == nil {
-		errorCopy.properties = make(map[Property]interface{}, 1)
-	} else {
-		errorCopy.properties = make(map[Property]interface{}, len(e.properties) + 1)
-		for k, v := range e.properties {
-			errorCopy.properties[k] = v
-		}
+// WithNewProperty adds a dynamic property to error instance if it were not already set.
```
Comment (PeterIvanov) [inline error.go]: were -> was
Or just "if it isn't already set".

File: error.go [hunk 3]
```diff
@@ -27,39 +27,45 @@ var _ fmt.Formatter = (*Error)(nil)
...
+// If an error already contained another value for the same property, it is NOT overwritten.
+func (e *Error) WithNewProperty(key Property, value interface{}) *Error {
```
Comment (PeterIvanov) [inline error.go]: I'm not sure the name is good enough. 
And even so, if this method may not do anything, shouldn't it return bool or error on property conflict?
Could you please describe some real life example for this method?

File: error.go [hunk 4]
```diff
@@ -27,39 +27,45 @@ var _ fmt.Formatter = (*Error)(nil)
...
-	errorCopy.underlying = newUnderying
-	return &errorCopy
+	if len(newUnderlying) == len(underlying) {
+		return e
+	}
+
+	l := len(newUnderlying)
+	newUnderlying = newUnderlying[:l:l]
+	errorCopy := e.WithProperty(propertyUnderlying, newUnderlying)
+	errorCopy.hasUnderlying = true
```
Comment (PeterIvanov) [inline error.go]:
```suggestion
	errorCopy.hasUnderlying = l > 0
```

File: id.go [hunk 5]
```diff
@@ -1,11 +1,23 @@
...
 
-var internalID int64
+var internalID uint64 = 0x123456789abcdef0
+var idMtx sync.Mutex
 
 // nextInternalID creates next unique id for errorx entities.
 // All equality comparison should take id into account, lest there be some false positive matches.
-func nextInternalID() int64 {
-	return atomic.AddInt64(&internalID, 1)
+func nextInternalID() uint64 {
```
Comment (PeterIvanov) [inline id.go]: A link to a paper describing these xorshift parameters would look good here. 
As this internal id **must** be unique, it is important to know exactly when to expect the first repetition. 

And since the current version of the code actually never uses the randomness in this id, is this change really needed?

File: property.go [hunk 6]
```diff
@@ -65,13 +69,41 @@ func ExtractProperty(err error, key Property) (interface{}, bool) {
...
+}
+
+// withProperty creates new map with property added or overwritten
+func (pm *propertyMap) with(p Property, value interface{}) *propertyMap {
+	return &propertyMap{p: p, value: value, next: pm}
+}
+
+func (pm *propertyMap) get(p Property) (value interface{}, ok bool) {
+	for pm != nil {
+		if pm.p == p {
```
Comment (PeterIvanov) [inline property.go]: Why do we rely on pointers equality here? Is it in any way better than to compare by id? If not, I would suggest we use id equality here, as it is less confusing and brittle.

File: property.go [hunk 7]
```diff
@@ -65,13 +69,41 @@ func ExtractProperty(err error, key Property) (interface{}, bool) {
...
+			id:    nextInternalID(),
+			label: label,
+		},
+	}
+	return p
+}
+
+// propertyMap represents map of properties.
+// Compared to builtin type, it uses less allocations and reallocations on copy.
+// It is simple binary search tree that relies on property id randomness for balancing.
```
Comment (PeterIvanov) [inline property.go]: This comment is false. Please revisit all the new comments and see if they are up to date.

File: property.go [hunk 8]
```diff
@@ -8,7 +8,11 @@ import (
 // Property value belongs to an error instance only, never inherited from a type.
 // Property visibility is hindered by Wrap, preserved by Decorate.
 type Property struct {
-	id    int64
+	*property
+}
+
+type property struct {
+	id    uint64
 	label string
```
Comment (PeterIvanov) [inline property.go]: It looks like we never use the label at this moment. If we remove it, the property key may be a simple unique id wrapper.

File: error.go (hunk 4 above)
Comment (funny-falcon-at-joomcode) [reply error.go]: No, because length could only grow, and not grown length == 0 were caught couple of lines above.

File: id.go (hunk 5 above)
Comment (funny-falcon-at-joomcode) [reply id.go]: No, it isn't needed, and I'm going to revert it.

ps: xorshift is a full period generator for 2^n-1 period.

File: property.go (hunk 6 above)
Comment (funny-falcon-at-joomcode) [reply property.go]: Equality by id is not better in any way, because pointer equality is just enough in this case.
Equality by pointer is a bit better for the case of one property because no need to fetch property id.

File: property.go (hunk 7 above)
Comment (funny-falcon-at-joomcode) [reply property.go]: Ok. Forgot about.

File: property.go (hunk 8 above)
Comment (funny-falcon-at-joomcode) [reply property.go]: I want to to add "info property" in next PR - property that will be pushed to Error message.
Such property will have to have name.

File: error.go (hunk 3 above)
Comment (funny-falcon-at-joomcode) [reply error.go]: > And even so, if this method may not do anything, shouldn't it return bool or error on property conflict?

In my use case, it just prevents double write of same value, therefore, it doesn't need to signal anything. And I want to save liquid interface.

> I'm not sure the name is good enough.

Which name do you suggest?

File: builder.go (hunk 1 above)
Comment (PeterIvanov) [reply builder.go]: Oh, I missed a benchmark in commit message. OK. 
Well, this is sad, actually. Go compiler fails to perform an escape analysis, so it is cheaper to risk extra copying.

File: error.go (hunk 4 above)
Comment (PeterIvanov) [reply error.go]: True. Still, it may not be cleat at the first glance, and some future change may accidentally break this, but that's a minor point. 
Ideally, the code should be as I suggested, and the compiler should eliminate the 'always true' calculation.

File: error.go [hunk 9]
```diff
@@ -25,41 +25,37 @@ var _ fmt.Formatter = (*Error)(nil)
 // If an error already contained another value for the same property, it is overwritten.
 // It is a caller's responsibility to accumulate and update a property, if needed.
 // Dynamic properties is a brittle mechanism and should therefore be used with care and in a simple and robust manner.
+// Currently, properties are implemented as a linked list, therefore it is not safe to have many dozens of them. But couple of dozen is just ok.
```
Comment (g7r) [inline error.go]: It is safe actually. Maybe just not as performant as it could be with map.

File: property.go (hunk 6 above)
Comment (PeterIvanov) [reply property.go]: OK, let's leave this as pointer equality check; please leave a comment to a private structure to clarify this point, and you can also remove the id from property altogether, as it is actually never used.

File: property.go (hunk 8 above)
Comment (PeterIvanov) [reply property.go]: Good, then we'll stick with pointer equality, see comment above.

File: error.go (hunk 3 above)
Comment (funny-falcon-at-joomcode) [reply error.go]: Ok, I decided to remove this method. In my use case, I will add private function.

File: error.go (hunk 9 above)
Comment (funny-falcon-at-joomcode) [reply error.go]: I never consider `O(n^2)` as a safe :-)

Comment (g7r) [reply error.go]: The documentation should be clear though. `WithProperty` will behave correctly even with hundred million dozen properties. But the performance wouldn't be pleasant. You should mention that `WithProperty` costs O(N) instead of declaring it as unsafe.
